### PR TITLE
Base classes for accelerator refactoring

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?
Adds the base classes from #5616

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

Co-Authored with @awaelchli

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
looking forward to it

**Review by SeanNaren — APPROVED** [review]
Looks good!

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+                dataloader_idx (int): The index of the dataloader that produced this batch
+                    (only if multiple val dataloaders used)
+        """
+        batch = self.to_device(args[0])
+
+        args[0] = batch
+
+        with self.precision_plugin.val_step_context():
+            with self.training_type_plugin.val_step_context():
+                return self.lightning_module.validation_step(*args)
```
Comment (SeanNaren) [inline pytorch_lightning/accelerators/accelerator.py:150]: This type of logic makes me think the precision plugin should live in the training type plugin, and that the base training type plugin manage the precision logic

Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:150]: Let's merge it as is and then change it later on :)

**Comment by Borda** [comment]
replacing usage of all unexisting Plugins and annotating with `fixme`

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    def __init__(self):
+        self._model = None
+        self._results = None
+        self.global_rank = 0
+
+    @property
+    @abstractmethod
+    def on_gpu(self) -> bool:
+        """Returns whether the current process is done on GPU"""
+        raise NotImplementedError
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:37]:
```suggestion
```
not needed since it is abstract

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    @property
+    @abstractmethod
+    def root_device(self) -> torch.device:
+        """Returns the root device"""
+        raise NotImplementedError
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:43]:
```suggestion
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    @abstractmethod
+    def model_to_device(self):
+        """Moves the model to the correct device"""
+        raise NotImplementedError
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:48]:
```suggestion
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    @property
+    @abstractmethod
+    def is_global_zero(self) -> bool:
+        """Whether the current process is the rank zero process not only on the local node, but for all nodes."""
+        raise NotImplementedError
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:54]:
```suggestion
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+    @abstractmethod
+    def reduce(self, output, *args, **kwargs):
+        """Reduces the given output (e.g. across GPUs/Processes)"""
+        raise NotImplementedError
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:59]:
```suggestion
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+    @abstractmethod
+    def barrier(self, name: Optional[str] = None):
+        """Forces all possibly joined processes to wait for each other"""
+        raise NotImplementedError
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:64]:
```suggestion
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    @abstractmethod
+    def broadcast(self, obj: object, src: int = 0) -> object:
+        """Broadcasts an object to all processes"""
+        raise NotImplementedError
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:69]:
```suggestion
```

**Review by Borda — APPROVED** [review]
for all these `@abstractmethod` with `raise NotImplementedError` use only one... 
- `@abstractmethod` cannot be instantiated till all these methods are implemented
- `raise NotImplementedError` can be instantiated, but will crash when touching one of these

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+import torch
+from torch.optim import Optimizer
+
+from pytorch_lightning.core import LightningModule
+from pytorch_lightning.plugins.base_plugin import Plugin
+
+
+class PrecisionPlugin(Plugin):
+    EPSILON = 1e-6
+    precision = 32
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/plugins/precision/precision_plugin.py:26]: Actually not. This is a class-constant. Not supposed to change at all

File: pytorch_lightning/accelerators/accelerator.py [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+        model = self.lightning_module
+        if model is not None:
+            return model.transfer_batch_to_device(batch, device)
+        return move_data_to_device(batch, device)
+
+    def on_train_start(self):
+        """Hook to do something upon the training start"""
+        pass
+
+    def training_step(self, args):
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:118]: I think this has to be without the asterisk. At least it was that way before and works...

File: pytorch_lightning/accelerators/accelerator.py [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+        self.connect_precision_plugin(self.precision_plugin)
+        self.optimizers = trainer.convert_to_lightning_optimizers(self.optimizers)
+
+    @property
+    def model(self) -> torch.nn.Module:
+        """Returns the model. This can also be a wrapped LightningModule.
+        For retrieving the pure LightningModule use :attr:`Accelerator.lightning_module`
+
+        """
+        return self.training_type_plugin.model
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:77]: Not sure. Actually there are many references to that in all parts of our code. So it's kind of public

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/plugins/precision/precision_plugin.py:26]: so capital?

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py:156]: agreed with @SeanNaren - that cleans up the interface for the accelerator then. conceptually the accelerator simply manages the training plugin with additional lifecycle hooks, and the training plugin is responsible for precision/whatever else(?)

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        """Connects this plugin to the accelerator and the training process"""
+        return model, optimizers, lr_schedulers
+
+    def backward(
+        self,
+        model: LightningModule,
+        closure_loss: torch.Tensor,
+        optimizer: torch.optim.Optimizer,
+        opt_idx: int,
+        should_accumulate: bool,
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/precision_plugin.py:47]: `should_accumulate` isn't used anywhere here?

File: pytorch_lightning/plugins/base_plugin.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import contextlib
+
+import torch
+
+
+class Plugin(object):
```
Comment (ananthsub) [inline pytorch_lightning/plugins/base_plugin.py:19]: are all of these needed for *all* plugins? are there any steps here that we could/should move to the training type plugin?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    def rpc_enabled(self) -> bool:
+        return False
+
+    def start_training(self, trainer: "Trainer") -> None:
+        # double dispatch to initiate the training loop
+        self._results = trainer.train()
+
+    def start_testing(self, trainer: "Trainer") -> None:
+        # double dispatch to initiate the test loop
+        self._results = trainer.run_test()
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:112]: this seems backwards to me - the plugin is the thing that initiates the training? i would've thought it was the other way around.

Trainer -> set up accelerator -> setup training type plugin -> train

**Review by ananthsub — COMMENTED** [review]
@justusschock  @awaelchli @SeanNaren what's the expected control flow across trainer/training loop/accelerator/training type plugin? i'm getting mixed up in trying to figure out what will call what when

**Comment by Borda** [comment]
@ananthsub good comments let address them in follow-up PRs :]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 14 above)
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:112]: Unfortunately it is not possible otherwise when we have to spawn training processes.

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 12 above)
Comment (justusschock) [reply pytorch_lightning/plugins/precision/precision_plugin.py:47]: For now this was leaved as I remembered there was an open PR using that. We definitely have to iterate after we merged all of these and do a final cleanup

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:156]: Yep, I also agree with @SeanNaren and there are plans to support that on top of the existing PoC since then this is probably easier to change then now.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+import torch
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.plugins.base_plugin import Plugin
+
+
+class TrainingTypePlugin(Plugin, ABC):
+    """A Plugin to change the behaviour of the training, validation and test-loop."""
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:26]: @justusschock - another question: currently [these training loop functions](https://github.com/PyTorchLightning/pytorch-lightning/blob/8b4fd44563b72c3351e0cc5c4a2c6a848d123cd4/pytorch_lightning/trainer/trainer.py#L496-L502) are set on the accelerator backend in `fit()` - should these be represented in the interface here or on the accelerator?  
What do you think about the training type plugin accepting these as constructor arguments, assuming we can define some base interfaces for train/evaluation/test loop?

Comment (justusschock) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:26]: I don't think this is needed anymore, since we pass the Trainer to start_training etc, where we can directly call trainer.train
